### ISSUE TITLE
Get inconsistent Contract storage data caused by stale/incorrect Snapshot

### ISSUE THREAD

**Issue by hsyodyssey** [scroll-tech/go-ethereum#37]
#### System information

Geth version: `geth version`
OS & Version: OSX
Commit hash : zkrollup-smt: ff17af3df89a1b3bc8c5593fc8c1d574aa0e16ab

#### Expected behaviour

Get the correct value of the persistent data in the contract.

#### Actual behaviour

The query result is not fully correct. Please see the description below for the details.

#### Steps to reproduce the behaviour

Here is the solidity code used in this experimental study.

```
// SPDX-License-Identifier: GPL-3.0

pragma solidity >=0.7.0 <0.9.0;

/**
 * @title Storage
 * @dev Store & retrieve value in a variable
 */
contract Storage {
    
    uint256 number;
    
    mapping(string => uint256) balances;

    function set_balance(uint256 num) public {
        number = num;
        balances["hsy"] = num;
        balances["ysh"] = num + 1;
    }
    
    function get_number() public view returns (uint256){
        return number;
    }

    function get_hsy() public view returns (uint256){
        return balances["hsy"];
    }

    function get_ysh() public view returns (uint256){
        return balances["ysh"];
    }
    
}
```

This contract has two variables, an uint256 variable named _number_, and a mapping variable named _balances_. The set_balance function will set the input value to number and balances["hsy"], and set the input value+1 to  balances["ysh"]. In addition, we have three get functions to query their values, respectively.

When we call the set_balance function with 100 as input, we expect the values of these variables to be: `number == 100, balances["hsy"] == 100, balances["ysh"] == 101`.

However, we can observe from the following figures that the query results are incorrect. Either all results are _100_ or all results are _101_.

![STORAGE AT OX962 18DDB (BLOC O x](https://user-images.githubusercontent.com/47173566/158019918-e72c81b8-e97f-4f3b-bc6d-c9483eec21d8.png)

![STORAGE AT OX962 18DDB (BLOC O](https://user-images.githubusercontent.com/47173566/158019925-f7f90310-8159-4a80-ab58-b932fa0e0a92.png)

#### Backtrace

TL;DR: This problem may be caused by the inconsistency between the **Snapshot** and the actual value.

---------------------------------------------------------------------------------------------------------

We call the set_balance function with 100 as an input. To locate the problem, I printed the relevant logs as follows. Each of the three variables we mentioned occupies a separate slot, so we can see the assignment of the three slots. And according to the following logs, it seems that the SetState function runs well and assgins the correct value to the corresponding slot.

```
INFO [03-12|21:52:49.772] Set State:                               0x0000000000000000000000000000000000000000000000000000000000000000=0x0000000000000000000000000000000000000000000000000000000000000064
INFO [03-12|21:52:49.774] Set State:                               0xbaded3bf529b04b554de2e4ee0f5702613335896b4041c50a5555b2d5e279f91=0x0000000000000000000000000000000000000000000000000000000000000064
INFO [03-12|21:52:49.777] Set State:                               0x37910dd9fc389b0cf2923975f06457bb9f45771422beae4687205a0988af00de=0x0000000000000000000000000000000000000000000000000000000000000065
```

But when the query functions are called we get a different value.  The value of _0x37910dd9fc389b0cf2923975f06457bb9f45771422beae4687205a0988af00de_ should be _**0x65**_ but gets _**0x64**_ instead.

```
INFO [03-12|22:01:58.959] Get State:                               0x0000000000000000000000000000000000000000000000000000000000000000=0x0000000000000000000000000000000000000000000000000000000000000064
INFO [03-12|22:01:59.675] Get State:                               0xbaded3bf529b04b554de2e4ee0f5702613335896b4041c50a5555b2d5e279f91=0x0000000000000000000000000000000000000000000000000000000000000064
INFO [03-12|22:02:00.457] Get State:                               0x37910dd9fc389b0cf2923975f06457bb9f45771422beae4687205a0988af00de=0x0000000000000000000000000000000000000000000000000000000000000064
```

This means that the problem is probably still caused by the getState function. 

We find that in the GetCommittedState function, the final value is actually obtained from the if segment `snap!=nil`. 

https://github.com/scroll-tech/go-ethereum/blob/zkrollup-smt/core/state/state_object.go#L229

```
	if s.db.snap != nil {
		if metrics.EnabledExpensive {
		if _, destructed := s.db.snapDestructs[s.addrHash]; destructed {
			return common.Hash{}
		}
		enc, err = s.db.snap.Storage(s.addrHash, crypto.Keccak256Hash(key.Bytes()))
	}
```

We can find in the stack at the breakpoint here that the values in snapshot are all 100 (0x64).

![image](https://user-images.githubusercontent.com/47173566/158021353-bb9edf09-97c8-43d0-92d9-55e9c8a25cf1.png)

I made an additional attempt to remove this _snap!=nil snippet_ and try to get the data from the trie each time. Then, we can get the correct data as the following firgure shows.

```
// Incorrect logic here. Only used for testing.
if s.db.snap != nil || s.db.snap == nil || err != nil {
if meter != nil {
	// If we already spent time checking the snapshot, account for it
	// and reset the readStart
	*meter += time.Since(readStart)
	readStart = time.Now()
}
if metrics.EnabledExpensive {
	meter = &s.db.StorageReads
}
if enc, err = s.getTrie(db).TryGet(key.Bytes()); err != nil {
	s.setError(err)
	return common.Hash{}
}
}
```
<img width="280" alt="STORAGE AT OX4B5 F85BO (BLOC" src="https://user-images.githubusercontent.com/47173566/158021549-5d19e08b-317c-40dc-96d3-eac75b0c99d1.png">

So the reason of this problem shall be the incorrect snapshot data. This problem will disappear after rebooting Geth(No Snapshot). But for a continuously running geth node, it is a problem that needs to be solved.

I found that this problem can be caused by the error in the `s.snapStorage` value at the time of calling stateDB `commit` function. https://github.com/scroll-tech/go-ethereum/blob/zkrollup-smt/core/state/statedb.go#L972

We can observe from the following logs that the sanpshot of all the Slots is updated with the same value (The logs below shows the hash of the Slot position). This is incorrect.

```
INFO [03-13|11:48:48.553] Update Snapshot                          0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563=0000000000000000000000000000000000000000000000000000000000000065
INFO [03-13|11:48:59.593] Update Snapshot                          0xa601d8e9cd2719ca27765dc16042655548d1ac3600a53ffc06b4a06a12b7c65c=0000000000000000000000000000000000000000000000000000000000000065
INFO [03-13|11:49:00.301] Update Snapshot                          0x43e5a6d45b6b646cf39869ca05026a8e8998e1cf7e9606beb8ca83e16c4a4e49=0000000000000000000000000000000000000000000000000000000000000065

```

The process of debugging in the middle is too long, so I'll omit it here and go straight to the conclusion:

I found this issue could be related to the `updateTrie` function in _state_object.go._. https://github.com/scroll-tech/go-ethereum/blob/zkrollup-smt/core/state/state_object.go#L322

The following is the function call stack when this problem is encountered.

```
     worker.go/commitTransactions->commitTransaction->commitTransaction
-> stete_processor.go/ApplyTransaction -> applyTransaction
-> stete_transasition.go/ApplyMessage -> TransitionDb
-> vm/evm.go/Call -> vm/interpreter.go/run -> vm/logger.go/CaptureStateAfter
-> statedb.go/GetStorageProof ->  StorageTrie
-> state_object.go/updateTrie
```

**Comment by lispc** [comment]
so you mean key-values in snapStorage are incorrect? 

the data flow is: StateObject.dirtyStorage->StateObject.pendingStorage->StateDb.snapStorage.   Just by reading codes i cannot find the reason. We will do more debugging.

**Comment by hsyodyssey** [comment]
> so you mean key-values in snapStorage are incorrect?
> 
> the data flow is: StateObject.dirtyStorage->StateObject.pendingStorage->StateDb.snapStorage. Just by reading codes i cannot find the reason. We will do more debugging.

Yes, the values in the snapStorage are incorret. This is indeed strange, and I also found that the process you mentioned looks fine. However, the temporary value of Snapshot was already wrong before statesb.Commit.

**Comment by hsyodyssey** [comment]
I tried to print the values of statedb.snapshptStorage before and after `calling cpy.updateTrie (s.db)`(https://github.com/scroll-tech/go-ethereum/blob/zkrollup-smt/core/state/statedb.go#L361). Here are the corresponding logs.

```
INFO [03-14|20:17:24.194] StorageTrie Before UpdateTrie            0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563=0000000000000000000000000000000000000000000000000000000000000005
INFO [03-14|20:17:24.194] StorageTrie Before UpdateTrie            0xa601d8e9cd2719ca27765dc16042655548d1ac3600a53ffc06b4a06a12b7c65c=0000000000000000000000000000000000000000000000000000000000000005

INFO [03-14|20:17:24.198] StorageTrie After UpdateTrie             0xa601d8e9cd2719ca27765dc16042655548d1ac3600a53ffc06b4a06a12b7c65c=0000000000000000000000000000000000000000000000000000000000000006
INFO [03-14|20:17:24.198] StorageTrie After UpdateTrie             0x43e5a6d45b6b646cf39869ca05026a8e8998e1cf7e9606beb8ca83e16c4a4e49=0000000000000000000000000000000000000000000000000000000000000006
INFO [03-14|20:17:24.198] StorageTrie After UpdateTrie             0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563=0000000000000000000000000000000000000000000000000000000000000006
```

**Comment by lispc** [comment]
i have a guess.... https://github.com/scroll-tech/go-ethereum/pull/33/commits/f73f9d0d96c3e687b254605462627d43eef0fe12

it is too strange for them to have same value... so my guess is ptr/shallow_copy...

**Comment by hsyodyssey** [comment]
> f73f9d0


Cool! It makes perfect sense! In the original code they used RLP to make a deep copy. https://github.com/ethereum/go-ethereum/blob/master/core/state/state_object.go#L350. The code just simply removes the RLP part.

```
v, _ = rlp.EncodeToBytes(common.TrimLeftZeroes(value[:]))
...
storage[crypto.HashData(hasher, key[:])] = v
```

**Comment by hsyodyssey** [comment]
I believe your guess is correct. The following logs illustrate that the positions of the `storage` elements are the same as the positions of the right side variable `vaule`.

```
[LOG] The Address of Storage 0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563 : 0xc00545e360 The Address of Value: 0xc00547eba0 
[LOG] The Address of Storage 0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563 : 0xc00547eba0 The Address of Value: 0xc005603120 
[LOG] The Address of Storage 0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563 : 0xc005603120 The Address of Value: 0xc005603120 
[LOG] The Address of Storage 0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563 : 0xc005603120 The Address of Value: 0xc005654920 
[LOG] The Address of Storage 0xa601d8e9cd2719ca27765dc16042655548d1ac3600a53ffc06b4a06a12b7c65c : 0xc005603120 The Address of Value: 0xc005654920 
[LOG] The Address of Storage 0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563 : 0xc005654920 The Address of Value: 0xc005654920 
[LOG] The Address of Storage 0xa601d8e9cd2719ca27765dc16042655548d1ac3600a53ffc06b4a06a12b7c65c : 0xc005603120 The Address of Value: 0xc005654920 
[LOG] The Address of Storage 0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563 : 0xc005654920 The Address of Value: 0xc00567dda0 
[LOG] The Address of Storage 0xa601d8e9cd2719ca27765dc16042655548d1ac3600a53ffc06b4a06a12b7c65c : 0xc005654920 The Address of Value: 0xc00567dda0 
[LOG] The Address of Storage 0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563 : 0xc00567dda0 The Address of Value: 0xc00567dda0 
[LOG] The Address of Storage 0xa601d8e9cd2719ca27765dc16042655548d1ac3600a53ffc06b4a06a12b7c65c : 0xc005654920 The Address of Value: 0xc00567dda0 
[LOG] The Address of Storage 0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563 : 0xc00567dda0 The Address of Value: 0xc00567dda0 
[LOG] The Address of Storage 0xa601d8e9cd2719ca27765dc16042655548d1ac3600a53ffc06b4a06a12b7c65c : 0xc00567dda0 The Address of Value: 0xc00567dda0 
[LOG] The Address of Storage 0xa601d8e9cd2719ca27765dc16042655548d1ac3600a53ffc06b4a06a12b7c65c : 0xc00567dda0 The Address of Value: 0xc0056e2e40 
[LOG] The Address of Storage 0x43e5a6d45b6b646cf39869ca05026a8e8998e1cf7e9606beb8ca83e16c4a4e49 : 0xc00567dda0 The Address of Value: 0xc0056e2e40 
[LOG] The Address of Storage 0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563 : 0xc00567dda0 The Address of Value: 0xc0056e2e40 
[LOG] The Address of Storage 0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563 : 0xc00567dda0 The Address of Value: 0xc0056e2e40 
[LOG] The Address of Storage 0xa601d8e9cd2719ca27765dc16042655548d1ac3600a53ffc06b4a06a12b7c65c : 0xc0056e2e40 The Address of Value: 0xc0056e2e40 
[LOG] The Address of Storage 0x43e5a6d45b6b646cf39869ca05026a8e8998e1cf7e9606beb8ca83e16c4a4e49 : 0xc00567dda0 The Address of Value: 0xc0056e2e40 
[LOG] The Address of Storage 0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563 : 0xc00567dda0 The Address of Value: 0xc0056e2e40 
[LOG] The Address of Storage 0xa601d8e9cd2719ca27765dc16042655548d1ac3600a53ffc06b4a06a12b7c65c : 0xc0056e2e40 The Address of Value: 0xc0056e2e40 
[LOG] The Address of Storage 0x43e5a6d45b6b646cf39869ca05026a8e8998e1cf7e9606beb8ca83e16c4a4e49 : 0xc0056e2e40 The Address of Value: 0xc0056e2e40 
```

**Comment by lightsing** [comment]
with snapshot disabled
![image](https://user-images.githubusercontent.com/15951701/158330187-3030b16e-c8da-42d0-9b36-0fa90ec132df.png)

**Comment by lightsing** [comment]
fixed in f73f9d0d96c3e687b254605462627d43eef0fe12
